src/core/lib/security/transport/security_handshaker.cc

```diff
@@ -227,7 +228,18 @@ static grpc_error *on_handshake_next_done_locked(
                        &h->on_handshake_data_received_from_peer);
     return error;
   }
+  // Handle TSI errors.
   if (result != TSI_OK) {
+    if (result == TSI_SEND_ALERT_AND_CLOSE && bytes_to_send_size > 0) {
+      gpr_log(GPR_DEBUG,
+              "Encountered TSI error. Sending alert message before closing");
+      grpc_slice to_send = grpc_slice_from_copied_buffer(
+          (const char *)bytes_to_send, bytes_to_send_size);
+      grpc_slice_buffer_reset_and_unref_internal(exec_ctx, &h->outgoing);
+      grpc_slice_buffer_add(&h->outgoing, to_send);
+      grpc_endpoint_write(exec_ctx, h->args->endpoint, &h->outgoing,
+                          &h->on_handshake_data_sent_to_peer);
+    }
     return grpc_set_tsi_error_result(
```

Comment:
In the case of TSI_SEND_ALERT_AND_CLOSE, if we return a grpc_error here, then the error will be propagated to do_handshaker_next_locked(), immediately after this, security_handshake_failed_locked and security_handshaker_unref will be called in Line 347 and Line 424. The security handshaker object could be destroyed before on_handshake_data_sent_to_peer callback is executed (since grpc_endpoint_write Line 240 is async).

Reply:
Thanks for pointing that out. I've updated it so that an error isn't returned in this case. The error is now created during the write callback so that the handshaker can safely be shut down after the write completes.